### PR TITLE
Unify C++ Human Pose Estimation demo

### PR DESCRIPTION
**Pull request by anzhella-pankratova**

### CODE REVIEW

File: demos/common/models/include/models/results.h [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include <map>
 //#include "metadata.h"
```
Comment (vladimir-dudnik) [inline demos/common/models/include/models/results.h:21]: if this line not needed please remove that

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: demos/common/models/src/hpe_model_openpose.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    if (pafsOutputDims.size() != 4 || pafsOutputDims[0] != 1 || pafsOutputDims[1] != 2 * (keypointsNumber + 1))
+        throw std::runtime_error("1x" + std::to_string(2 * (keypointsNumber + 1)) + "xHFMxWFM dimension of model's output is expected");
+    const SizeVector& heatmapsOutputDims = (*outputIt++).second->getTensorDesc().getDims();
+    if (heatmapsOutputDims.size() != 4 || heatmapsOutputDims[0] != 1 || heatmapsOutputDims[1] != keypointsNumber + 1)
+        throw std::runtime_error("1x" + std::to_string(keypointsNumber + 1) + "xHFMxWFM dimension of model's heatmap is expected");
+    if (pafsOutputDims[2] != heatmapsOutputDims[2] || pafsOutputDims[3] != heatmapsOutputDims[3])
+        throw std::runtime_error("output and heatmap are expected to have matching last two dimensions");
+}
+
+cv::Size HPEOpenPose::reshape(InferenceEngine::CNNNetwork& cnnNetwork, int targetSize) {
```
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/common/models/src/hpe_model_openpose.cpp]: I believe aspect ratio should be provided to the model instead of actual image size: currently a constructor takes two values for height: one is `targetSize`, the second is ` cv::Size::height`. That is ambiguous.

File: demos/common/models/src/hpe_model_openpose.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    *static_cast<ResultBase*>(result) = static_cast<ResultBase&>(infResult);
+
+    auto outputMapped = infResult.outputsData[outputsNames[0]];
+    auto heatMapsMapped = infResult.outputsData[outputsNames[1]];
+
+    const SizeVector& outputDims = outputMapped->getTensorDesc().getDims();
+    const SizeVector& heatMapDims = heatMapsMapped->getTensorDesc().getDims();
+
+    const float *predictions = outputMapped->rmap().as<float*>();
+    const float *heats = heatMapsMapped->rmap().as<float*>();
```
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/common/models/src/hpe_model_openpose.cpp]: Please remove `reinterpret_cast` and `const_cast` below.

File: demos/common/pipelines/src/async_pipeline.cpp [hunk 4]
```diff
@@ -51,6 +52,10 @@ AsyncPipeline::AsyncPipeline(std::unique_ptr<ModelBase>&& modelInstance, const C
     auto shapes = cnnNetwork.getInputShapes();
     for (auto& shape : shapes) {
         shape.second[0] = 1;
+        if (reshape != cv::Size()) {
```
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/common/pipelines/src/async_pipeline.cpp]: `ModelBase`'s `reshape()` has everything to perform reshaping on it's own without going to `AsyncPipeline`. `cnnNetwork.reshape(shapes);` should be removed from `AsyncPipeline`. Instead a common implementation in `ModelBase` or in each particular net should be introduced.

File: demos/common/models/src/hpe_model_openpose.cpp [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+bool HPEOpenPose::reshape(InferenceEngine::CNNNetwork& cnnNetwork) {
+    ICNNNetwork::InputShapes inputShapes = cnnNetwork.getInputShapes();
+    SizeVector& InputLayerDims = inputShapes.begin()->second;
+    double aspectRatio = inputImageSize.width / static_cast<double>(inputImageSize.height);
+    if (!targetSize) {
+        targetSize = InputLayerDims[2];
+    }
+    int inputWidth = static_cast<int>(std::round(targetSize * aspectRatio));
+    int height = static_cast<int>((inputWidth + stride - 1) / stride) * stride;
+    int width = static_cast<int>((targetSize + stride - 1) / stride) * stride;
```
Comment (Wovchena) [inline demos/common/models/src/hpe_model_openpose.cpp:85]: It looks `height` and `width` are mixed: `height` should depend on `targetSize` directly and `width` should depend through `aspectRatio` following help in python version.

File: demos/common/models/src/hpe_model_openpose.cpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    }
+    resizeFeatureMaps(pafs);
+
+    std::vector<HumanPose> poses = extractPoses(heatMaps, pafs);
+
+    cv::Size fullFeatureMapSize = heatMaps[0].size() * stride / upsampleRatio;
+    float scaleX = internalData.inputImgWidth /
+            static_cast<float>(fullFeatureMapSize.width - pad(1) - pad(3));
+    float scaleY = internalData.inputImgHeight /
+            static_cast<float>(fullFeatureMapSize.height - pad(0) - pad(2));
```
Comment (Wovchena) [inline demos/common/models/src/hpe_model_openpose.cpp:152]: It looks it's incorrect to have padding affecting scale values. `scaleX` and `scaleY` correspond to scale introduced by the net during inference. That scaling doesn't care if the pixels represent an actual image or they are just padded. That removes the need in `pad` in postprocessing, so you shouldn't store it in `HPEOpenPose`.

File: demos/human_pose_estimation_demo/main.cpp [hunk 7]
```diff
@@ -44,186 +125,189 @@ bool ParseAndCheckCommandLine(int argc, char* argv[]) {
...
-            estimator.waitCurr();
-            t1 = std::chrono::high_resolution_clock::now();
-            ms detection = std::chrono::duration_cast<ms>(t1 - t0);
-            t0 = std::chrono::high_resolution_clock::now();
-            ms wall = std::chrono::duration_cast<ms>(t0 - wallclock);
-            wallclock = t0;
+        //------------------------------- Preparing Input ------------------------------------------------------
+        slog::info << "Reading input" << slog::endl;
+        auto startTime = std::chrono::steady_clock::now();
+        auto cap = openImagesCapture(FLAGS_i, FLAGS_loop);
```
Comment (Wovchena) [inline demos/human_pose_estimation_demo/main.cpp:212]:
```suggestion
        auto cap = openImagesCapture(FLAGS_i, FLAGS_loop);
        auto startTime = std::chrono::steady_clock::now();
```

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: demos/common/models/src/hpe_model_openpose.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (anzhella-pankratova) [reply demos/common/models/src/hpe_model_openpose.cpp]: Done.

File: demos/common/models/src/hpe_model_openpose.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (anzhella-pankratova) [reply demos/common/models/src/hpe_model_openpose.cpp]: Done.

File: demos/common/pipelines/src/async_pipeline.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (anzhella-pankratova) [reply demos/common/pipelines/src/async_pipeline.cpp]: Done.

File: demos/common/models/src/hpe_model_openpose.cpp (hunk 5 above)
Comment (anzhella-pankratova) [reply demos/common/models/src/hpe_model_openpose.cpp]: Done.

File: demos/common/models/src/hpe_model_openpose.cpp (hunk 6 above)
Comment (anzhella-pankratova) [reply demos/common/models/src/hpe_model_openpose.cpp]: Done.

File: demos/human_pose_estimation_demo/main.cpp (hunk 7 above)
Comment (anzhella-pankratova) [reply demos/human_pose_estimation_demo/main.cpp]: Done.

**Review by Wovchena — APPROVED** [review]
LGTM. Are you going to add other models support in a separate PR?

**Comment by anzhella-pankratova** [comment]
> LGTM. Are you going to add other models support in a separate PR?

I think yes, let it be in a separate PR.

**Review by vladimir-dudnik — APPROVED** [review]

File: demos/common/models/src/openpose_decoder.cpp [hunk 8]
```diff
@@ -1,16 +1,27 @@
-// Copyright (C) 2018-2019 Intel Corporation
-// SPDX-License-Identifier: Apache-2.0
+/*
+// Copyright (C) 2018-2020 Intel Corporation
```
Comment (asutic) [inline demos/common/models/src/openpose_decoder.cpp:2]: We are in 2021 now. Other files require the same update.

File: demos/human_pose_estimation_demo/README.md [hunk 9]
```diff
@@ -29,16 +29,24 @@ InferenceEngine:
...
-    -m "<path>"                Required. Path to the Human Pose Estimation model (.xml) file.
-    -d "<device>"              Optional. Specify the target device for Human Pose Estimation (the list of available devices is shown below). Default value is CPU. Use "-d HETERO:<comma-separated_devices_list>" format to specify HETERO plugin. The application looks for a suitable plugin for the specified device.
-    -pc                        Optional. Enable per-layer performance report.
-    -no_show                   Optional. Do not show processed video.
-    -black                     Optional. Show black background.
-    -r                         Optional. Output inference results as raw values.
-    -u                         Optional. List of monitors to show initially.
+    -h                        Print a usage message.
+    -at "<type>"              Required. Type of the network, either "ae" for
+                              Associative Embedding or "openpose" for OpenPose.
```
Comment (asutic) [inline demos/human_pose_estimation_demo/README.md:34]: why not just 'op'?

File: demos/common/models/src/openpose_decoder.cpp [hunk 10]
```diff
@@ -1,16 +1,27 @@
...
 #include <utility>
 #include <vector>
 
 #include <samples/common.hpp>
+#include "models/openpose_decoder.h"
 
-#include "peak.hpp"
 
-namespace human_pose_estimation {
 Peak::Peak(const int id, const cv::Point2f& pos, const float score)
```
Comment (asutic) [inline demos/common/models/src/openpose_decoder.cpp:25]: Does it make sense to write a G-API kernel for finding peaks?

Comment (vladimir-dudnik) [reply demos/common/models/src/openpose_decoder.cpp:25]: potentially yes, but this will require series of experiments, so better to have a separate PR for such work

File: demos/common/models/src/openpose_decoder.cpp (hunk 8 above)
Comment (anzhella-pankratova) [reply demos/common/models/src/openpose_decoder.cpp]: Updated.

File: demos/human_pose_estimation_demo/README.md (hunk 9 above)
Comment (anzhella-pankratova) [reply demos/human_pose_estimation_demo/README.md:34]: The "openpose" is not so long as "associative embedding", therefore I think the word reduction is not necessary here.

File: demos/common/models/src/openpose_decoder.cpp (hunk 10 above)
Comment (asutic) [reply demos/common/models/src/openpose_decoder.cpp:25]: I agree. It is certainly better to keep it in separate PR. But, thanks for the confirmation!

File: demos/common/cpp/models/include/models/hpe_model_openpose.h [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    void prepareInputsOutputs(InferenceEngine::CNNNetwork & cnnNetwork) override;
+
+    const int minJointsNumber = 3;
+    const int stride = 8;
+    const cv::Vec3f meanPixel = cv::Vec3f::all(128);
+    const float minPeaksDistance = 3.0f;
+    const float midPointsScoreThreshold = 0.05f;
+    const float foundMidPointsRatioThreshold = 0.8f;
+    const float minSubsetScore = 0.2f;
+    const int upsampleRatio = 4;
```
Comment (IRDonch) [inline demos/common/cpp/models/include/models/hpe_model_openpose.h:44]: Seems like these could be `static`.

File: demos/tests/data_sequences.py [hunk 12]
```diff
@@ -88,7 +88,6 @@
     ],
 
     'human-pose-estimation': [
-        image_net_arg('00000002'),
```
Comment (IRDonch) [inline demos/tests/data_sequences.py:91]: What's the purpose of this change?

**Review by IRDonch — CHANGES_REQUESTED** [review]

File: demos/common/cpp/models/include/models/hpe_model_openpose.h (hunk 11 above)
Comment (anzhella-pankratova) [reply demos/common/cpp/models/include/models/hpe_model_openpose.h]: Updated

File: demos/tests/data_sequences.py (hunk 12 above)
Comment (anzhella-pankratova) [reply demos/tests/data_sequences.py:91]: This sequence consists of images with incorrect aspect ratio, which does not fit the model shape, so the error during model preprocessing occurs.

Comment (IRDonch) [reply demos/tests/data_sequences.py:91]: OK.

File: demos/common/cpp/models/include/models/hpe_model_openpose.h (hunk 11 above)
Comment (IRDonch) [reply demos/common/cpp/models/include/models/hpe_model_openpose.h]: You updated the integers, but what about the rest of the constants?

Comment (anzhella-pankratova) [reply demos/common/cpp/models/include/models/hpe_model_openpose.h]: Updated them too.

**Review by IRDonch — APPROVED** [review]
